### ISSUE TITLE
a broken query stops dnstap_receiver to work.

### ISSUE THREAD

**Issue by motok** [dmachard/dnstap-receiver#35]
- dnstap_reciever stops working when receiving a dns query with (qdcount>0) and (no qtype&qclass), which is broken as a query but some abuser sends. Please take a look at the following error message.
```
Task exception was never retrieved
future: <Task finished coro=<cb_onconnect() done, defined at /path/to/dnstap-receiver/dnstap_receiver/inputs/input_socket.py:13> exception=error('unpack requires a buffer of 4 bytes')>
Traceback (most recent call last):
  File "/path/to/dnstap-receiver/dnstap_receiver/inputs/input_socket.py", line 67, in cb_onconnect
    await dnstap_decoder.cb_ondnstap(dnstap_protobuf, payload, cfg, queues_list, stats, geoip_reader, cache)
  File "/path/to/dnstap-receiver/dnstap_receiver/inputs/dnstap_decoder.py", line 128, in cb_ondnstap
    qname, qtype = dns_parser.decode_question(dns_payload)
  File "/path/to/dnstap-receiver/dnstap_receiver/dns/parser.py", line 20, in decode_question
    qtype, qclass  = struct.unpack('!HH', buf[1:5])
struct.error: unpack requires a buffer of 4 bytes
```
- This dnstap_receiver locates behind nsd with the following dnstap configuration.
```
dnstap:
        dnstap-enable: yes
        dnstap-socket-path: "/var/run/nsd/dnstap.sock"
        dnstap-send-identity: no
        dnstap-send-version: no
        dnstap-log-auth-query-messages: yes
```
- and invoking command line is as follows.
```
dnstap_receiver -v -u /var/run/nsd/dnstap.sock -c my.conf
```
```
[my.conf]
filter:
  qname-regex: "[a-z0-9]{63}.example.org."
output:
  stdout:
    enable: true
    format: json
web-api:
  enable: false
```
- We might need to catch the task exception. My quick hack follows. (not tested enough)
```
$ diff -u dnstap_decoder.py.orig dnstap_decoder.py
--- dnstap_decoder.py.orig	2021-06-29 23:32:38.831166000 +0900
+++ dnstap_decoder.py	2021-07-01 16:30:16.538717000 +0900
@@ -125,10 +125,14 @@
 
     # common params
     if dns_qdcount:
-        qname, qtype = dns_parser.decode_question(dns_payload)
-        tap["qname"] = qname.decode(errors="ignore")
-        tap["rrtype"] = dns_rdatatypes.RDATATYPES[qtype]
-        
+        try:
+            qname, qtype = dns_parser.decode_question(dns_payload)
+        except:
+            clogger.warn('Dnstap decorder - no qname/qtype though dns_qdcount > 0: %s' % dns_payload)
+        else:
+            tap["qname"] = qname.decode(errors="ignore")
+            tap["rrtype"] = dns_rdatatypes.RDATATYPES[qtype]
+
     tap["rcode"] = dns_rcodes.RCODES[dns_rcode]
     tap["id"] = dns_id
 
```
- Thank you very much for your developing such a nice software. It helps me a lot.

**Comment by dmachard** [comment]
thanks to report this issue.
I pushed a commit to be more permissive during the decoding.

**Comment by motok** [comment]
Thank you very much for your patch. I am running new version and will watch what will happen.
btw, here are real world examples of the abusing dns messages just for your information.
```
2021-07-03 20:06:30,062 WARNING Dnstap decorder - no qname/qtype though dns_qdcount > 0: b'\x00\x00)%(\x00\x00\x00\x00\x00\x00\x00'
2021-07-04 02:14:28,019 WARNING Dnstap decorder - no qname/qtype though dns_qdcount > 0: b'\xdd\x11\x01 \x00\x01\x00\x00\x00\x00\x00\x00\x07version\x04bind\xc0\x1b\x00\x10\x00\x03'
```

**Comment by motok** [comment]
Thank you very much and I think the problem has been fixed by your patch.

**Comment by dmachard** [comment]
Fix included in the new v4.4.0 release.
